context/config_file.go

```diff
@@ -35,32 +31,126 @@ func parseConfigFile(fn string) (*configEntry, error) {
-func ParseDefaultConfig() (*configEntry, error) {
+func ParseDefaultConfig() (*Config, error) {
 	return parseConfigFile(configFile())
 }
 
-func parseConfig(r io.Reader) (*configEntry, error) {
+type AuthConfig struct {
+	User  string
+	Token string
+}
+
+type HostConfig struct {
+	Host  string
+	Auths []*AuthConfig
+}
+
+type Config struct {
+	Root     *yaml.Node
+	Hosts    []*HostConfig
+	Editor   string
+	Protocol string
+}
+
+func (c *Config) ConfigForHost(hostname string) (*HostConfig, error) {
+	for _, hc := range c.Hosts {
+		if hc.Host == hostname {
+			return hc, nil
+		}
+	}
+	return nil, fmt.Errorf("could not find config entry for \"%s\"", hostname)
+}
+
+func (c *Config) DefaultHostConfig() (*HostConfig, error) {
+	return c.ConfigForHost(defaultHostname)
+}
+
+func defaultConfig() Config {
+	return Config{
+		Protocol: "https",
+		// we leave editor as empty string to signal that we should use environment variables
+	}
+}
+
+func parseConfig(r io.Reader) (*Config, error) {
 	data, err := ioutil.ReadAll(r)
 	if err != nil {
 		return nil, err
 	}
-	var config yaml.Node
-	err = yaml.Unmarshal(data, &config)
+	var root yaml.Node
+	err = yaml.Unmarshal(data, &root)
 	if err != nil {
 		return nil, err
 	}
-	if len(config.Content) < 1 {
+	if len(root.Content) < 1 {
 		return nil, fmt.Errorf("malformed config")
 	}
-	for i := 0; i < len(config.Content[0].Content)-1; i = i + 2 {
-		if config.Content[0].Content[i].Value == defaultHostname {
-			var entries []configEntry
-			err = config.Content[0].Content[i+1].Decode(&entries)
-			if err != nil {
-				return nil, err
+	if root.Content[0].Kind != yaml.MappingNode {
+		return nil, fmt.Errorf("expected a top level map")
+	}
+
+	config := defaultConfig()
+	config.Root = &root
+
+	// TODO
+
+	// - [x] make hosts: nesting format work with new parsing code and config struct
+	// - [ ] support setting editor (or protocol)
+	// - [ ] implement new commands
+	// - [ ] migration code for old (non-hosts) config
+
+	topLevelKeys := root.Content[0].Content
+
+	for i, v := range topLevelKeys {
+		switch v.Value {
+		case "hosts":
+			// hosts is a map of hostname -> arrays of user auth configs
+			for j, v := range topLevelKeys[i+1].Content {
+				if v.Value == "" {
+					continue
+				}
+				hostConfig := HostConfig{}
+				hostConfig.Host = v.Value
+				authsRoot := topLevelKeys[i+1].Content[j+1]
+				// Each one of these is a map that holds user/oauth_token
+				for _, v := range authsRoot.Content {
+					authConfig := AuthConfig{}
+					authRoot := v.Content
+					// This is a map of user/token values
+					for y, v := range authRoot {
+						switch v.Value {
+						case "user":
+							authConfig.User = authRoot[y+1].Value
+						case "oauth_token":
+							authConfig.Token = authRoot[y+1].Value
+						}
+					}
+					hostConfig.Auths = append(hostConfig.Auths, &authConfig)
+				}
+				config.Hosts = append(config.Hosts, &hostConfig)
 			}
-			return &entries[0], nil
+		case "protocol":
```

Comment:
I wonder if we could get away with simply storing/parsing any key-value pair and letting another mechanism query the values and decide what to do with them?

I feel as if this should be a lower-level config mechanism which doesn't necessarily understand nor care about the meaning of any configuration options, apart from parsing per-host config, and then have another configuration layer that wraps the storage layer and understands what `protocol`, `ssh`, `editor` and such means.

Reply:
Could we be more specific with what "protocol" refers to here, for example by renaming the key to something like `clone_protocol` or `git_protocol`?

The reason is that for GitHub enterprise, we might need to support GHE API access over the `http:` protocol (some intranet installs don't have GHE accessible under SSL and that's okay I guess since it's intranet). To support that case, there might be an `api_protocol` or similarly named key under host-specific configuration, and therefore it might be generally a good idea to avoid vagueness when using the world "protocol" unqualified

Let me know if this makes sense; just thinking ahead here. https://github.com/github/hub/issues/2128#issuecomment-488694176

Reply:
> I wonder if we could get away with simply storing/parsing any key-value pair and letting another mechanism query the values and decide what to do with them?

I was considering an approach like this:

- continue store the root `*yaml.Node` in the `Config` struct. We need it to properly write the config back out in case it's changed via a `gh config...` invocation.
- add methods to the `Config` struct that lazily parse a desired config key instead of trying to parse it all up front. Each method is responsible for traversing the root and finding the desired node and parsing it. the `Hosts` method would need to be involved, but the simple k/v settings can share a helper.
- support a mock `Config` in tests (like we do with `Context`) whose various methods have predictable responses
- unit test the yaml parsing on its own (like we already do)

This way as we add new supported config values we add new methods to `Config`.

This approach means we can progressively parse the yaml and not fail because of an unknown key or malformed key that isn't relevant to the current invocation of `gh`.